File: apps/ios/QianwenApp/QianwenApiClient.swift
import Foundation

enum QianwenApiError: Error, LocalizedError {
    case invalidResponse
    case httpStatus(Int)

    var errorDescription: String? {
        switch self {
        case .invalidResponse:
            return "服务端响应无效。"
        case .httpStatus(let status):
            return "HTTP \(status)"
        }
    }
}

final class QianwenApiClient {
    private let baseURL: URL
    private let session: URLSession
    private let decoder = JSONDecoder()
    private let encoder = JSONEncoder()

    init(baseURL: URL = URL(string: "http://localhost:8787")!, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.session = session
    }

    func health() async throws -> HealthResponse {
        try await request(path: "/health", method: "GET")
    }

    func listConversations() async throws -> ConversationsResponse {
        try await request(path: "/conversations", method: "GET")
    }

    func createConversation(title: String) async throws -> CreateConversationResponse {
        try await request(path: "/conversations", method: "POST", body: CreateConversationRequest(title: title))
    }

    func updateConversation(conversationId: String, title: String? = nil, pinned: Bool? = nil) async throws -> CreateConversationResponse {
        try await request(
            path: "/conversations/\(conversationId)",
            method: "PATCH",
            body: UpdateConversationRequest(title: title, pinned: pinned)
        )
    }

    func deleteConversation(conversationId: String) async throws -> ConversationsResponse {
        try await request(path: "/conversations/\(conversationId)", method: "DELETE")
    }

    func getMessages(conversationId: String) async throws -> MessagesResponse {
        try await request(path: "/conversations/\(conversationId)/messages", method: "GET")
    }

    func clearMessages(conversationId: String) async throws -> MessagesResponse {
        try await request(path: "/conversations/\(conversationId)/messages", method: "DELETE")
    }

    func chat(conversationId: String, message: String) async throws -> ChatResponse {
        try await request(path: "/chat", method: "POST", body: ChatRequest(conversationId: conversationId, message: message))
    }

    func streamChat(
        conversationId: String,
        message: String,
        onEvent: @escaping @Sendable (ChatStreamEvent) async -> Void
    ) async throws {
        var request = URLRequest(url: baseURL.appendingPathComponent("chat/stream"))
        request.httpMethod = "POST"
        request.setValue("text/event-stream", forHTTPHeaderField: "accept")
        request.setValue("application/json", forHTTPHeaderField: "content-type")
        request.httpBody = try encoder.encode(ChatRequest(conversationId: conversationId, message: message))
        let (bytes, response) = try await session.bytes(for: request)
        try validate(response)

        let parser = ChatSseParser(decoder: decoder)

        for try await line in bytes.lines {
            if let event = try parser.parseLine(line) {
                await onEvent(event)
            }
        }
        if let event = try parser.flush() {
            await onEvent(event)
        }
    }

    private func request<T: Decodable>(path: String, method: String) async throws -> T {
        var request = URLRequest(url: baseURL.appendingPathComponent(String(path.dropFirst())))
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "content-type")
        let (data, response) = try await session.data(for: request)
        try validate(response)
        return try decoder.decode(T.self, from: data)
    }

    private func request<T: Decodable, B: Encodable>(path: String, method: String, body: B) async throws -> T {
        var request = URLRequest(url: baseURL.appendingPathComponent(String(path.dropFirst())))
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "content-type")
        request.httpBody = try encoder.encode(body)
        let (data, response) = try await session.data(for: request)
        try validate(response)
        return try decoder.decode(T.self, from: data)
    }

    private func validate(_ response: URLResponse) throws {
        guard let http = response as? HTTPURLResponse else {
            throw QianwenApiError.invalidResponse
        }
        guard (200..<300).contains(http.statusCode) else {
            throw QianwenApiError.httpStatus(http.statusCode)
        }
    }

}

final class ChatSseParser {
    private var eventName: String?
    private var dataLines: [String] = []
    private let decoder: JSONDecoder

    init(decoder: JSONDecoder = JSONDecoder()) {
        self.decoder = decoder
    }

    func parseLine(_ line: String) throws -> ChatStreamEvent? {
        if line.isEmpty {
            return try flush()
        }

        if line.hasPrefix("event:") {
            eventName = String(line.dropFirst("event:".count)).trimmingCharacters(in: .whitespaces)
            return nil
        }

        if line.hasPrefix("data:") {
            var data = String(line.dropFirst("data:".count))
            if data.hasPrefix(" ") {
                data.removeFirst()
            }
            dataLines.append(data)
            return nil
        }

        return nil
    }

    func flush() throws -> ChatStreamEvent? {
        guard !dataLines.isEmpty else {
            eventName = nil
            return nil
        }

        let payload = dataLines.joined(separator: "\n")
        let data = Data(payload.utf8)
        let eventType = (try? JSONSerialization.jsonObject(with: data) as? [String: Any])?["type"] as? String
        let type = eventType ?? eventName

        eventName = nil
        dataLines.removeAll(keepingCapacity: true)

        switch type {
        case "conversation":
            return .conversation(try decoder.decode(StreamConversationEvent.self, from: data))
        case "message":
            return .message(try decoder.decode(StreamMessageEvent.self, from: data))
        case "delta":
            return .delta(try decoder.decode(StreamDeltaEvent.self, from: data))
        case "done":
            return .done(try decoder.decode(StreamDoneEvent.self, from: data))
        case "error":
            return .error(try decoder.decode(StreamErrorEvent.self, from: data))
        default:
            return nil
        }
    }
}

File: apps/android/app/src/main/java/com/qianwen/demo/data/QianwenApiClient.kt
package com.qianwen.demo.data

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.withContext
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

@OptIn(ExperimentalSerializationApi::class)
class QianwenApiClient(
    private val baseUrl: String,
    private val client: OkHttpClient = OkHttpClient(),
    private val json: Json = Json {
        ignoreUnknownKeys = true
        explicitNulls = false
    }
) {
    private val contentType = "application/json; charset=utf-8".toMediaType()

    suspend fun health(): HealthResponse = get("/health")

    suspend fun listConversations(): ConversationsResponse = get("/conversations")

    suspend fun createConversation(title: String): CreateConversationResponse {
        return post("/conversations", CreateConversationRequest(title))
    }

    suspend fun updateConversation(
        conversationId: String,
        title: String? = null,
        pinned: Boolean? = null
    ): CreateConversationResponse {
        return patch("/conversations/$conversationId", UpdateConversationRequest(title, pinned))
    }

    suspend fun deleteConversation(conversationId: String): ConversationsResponse {
        return delete("/conversations/$conversationId")
    }

    suspend fun getMessages(conversationId: String): MessagesResponse {
        return get("/conversations/$conversationId/messages")
    }

    suspend fun clearMessages(conversationId: String): MessagesResponse {
        return delete("/conversations/$conversationId/messages")
    }

    suspend fun chat(conversationId: String, message: String): ChatResponse {
        return post("/chat", ChatRequest(conversationId, message))
    }

    suspend fun streamChat(
        conversationId: String,
        message: String,
        onEvent: suspend (ChatStreamEvent) -> Unit
    ) {
        val request = Request.Builder()
            .url("$baseUrl/chat/stream")
            .header("accept", "text/event-stream")
            .post(json.encodeToString(ChatRequest(conversationId, message)).toRequestBody(contentType))
            .build()
        val call = client.newCall(request)

        withContext(Dispatchers.IO) {
            val cancellationHandle = currentCoroutineContext()[Job]?.invokeOnCompletion { cause ->
                if (cause is CancellationException) {
                    call.cancel()
                }
            }

            try {
                call.execute().use { httpResponse ->
                    if (!httpResponse.isSuccessful) {
                        val detail = httpResponse.body?.string().orEmpty()
                        error("HTTP ${httpResponse.code}: $detail")
                    }

                    val source = httpResponse.body?.source() ?: error("Empty stream response body")
                    val parser = ChatSseParser(json)
                    var terminalReceived = false
                    while (!terminalReceived) {
                        currentCoroutineContext().ensureActive()
                        val line = source.readUtf8Line() ?: break
                        val event = parser.parseLine(line)
                        if (event != null) {
                            onEvent(event)
                            terminalReceived = event is StreamDoneEvent || event is StreamErrorEvent
                        }
                    }

                    if (!terminalReceived) {
                        val event = parser.flush()
                        if (event != null) {
                            onEvent(event)
                            terminalReceived = event is StreamDoneEvent || event is StreamErrorEvent
                        }
                    }
                    if (!terminalReceived) {
                        error("SSE 连接中断，未收到 done/error 事件。")
                    }
                }
            } finally {
                cancellationHandle?.dispose()
            }
        }
    }

    private suspend inline fun <reified T> get(path: String): T = request(path, "GET")

    private suspend inline fun <reified T> delete(path: String): T = request(path, "DELETE")

    private suspend inline fun <reified T, reified B> post(path: String, body: B): T {
        return request(path, "POST", json.encodeToString(body).toRequestBody(contentType))
    }

    private suspend inline fun <reified T, reified B> patch(path: String, body: B): T {
        return request(path, "PATCH", json.encodeToString(body).toRequestBody(contentType))
    }

    private suspend inline fun <reified T> request(path: String, method: String, body: okhttp3.RequestBody? = null): T {
        return withContext(Dispatchers.IO) {
            val request = Request.Builder()
                .url("$baseUrl$path")
                .method(method, body)
                .header("content-type", "application/json")
                .build()
            client.newCall(request).execute().use { response ->
                val responseText = response.body?.string().orEmpty()
                if (!response.isSuccessful) {
                    error("HTTP ${response.code}: $responseText")
                }
                json.decodeFromString(responseText)
            }
        }
    }
}
